source/extensions/filters/common/local_ratelimit/provider_singleton.cc

```diff
@@ -0,0 +1,199 @@
+  }
+
+  if (init_target_) {
+    init_target_->ready();
+  }
+}
+
+void RateLimiterProviderSingleton::TokenBucketSubscription::handleRemovedResource(
+    absl::string_view) {
+  // We simply reset the config and limiter here. The existing rate limiter will
+  // continue to work before the new config is received.
+  config_.reset();
+  token_bucket_config_hash_ = 0;
+  limiter_.reset();
+  init_target_ = nullptr;
+
+  // Reset the init target as we are now waiting for a new resource.
+  for (auto& [key, setter] : setters_) {
+    setter(parent_.fallback_always_deny_limiter_);
+  }
+}
+
+absl::Status RateLimiterProviderSingleton::TokenBucketSubscription::onConfigUpdate(
+    const std::vector<Config::DecodedResourceRef>& resources, const std::string&) {
+  ENVOY_BUG(resources.size() == 1,
+            fmt::format("for singleton resource subscription, only one resource should be "
+                        "added or removed at a time but got {}",
+                        resources.size()));
+  ENVOY_BUG(resources[0].get().name() == resource_name_,
+            fmt::format("for singleton resource subscription, the added resource name "
+                        "should be the same as the subscription resource name but got "
+                        "{} != {}",
+                        resources[0].get().name(), resource_name_));
+
+  handleAddedResource(resources[0]);
+  return absl::OkStatus();
+}
+
+absl::Status RateLimiterProviderSingleton::TokenBucketSubscription::onConfigUpdate(
+    const std::vector<Config::DecodedResourceRef>& added_resources,
+    const Protobuf::RepeatedPtrField<std::string>& removed_resources, const std::string&) {
+  ENVOY_BUG(added_resources.size() + removed_resources.size() == 1,
+            fmt::format("for singleton resource subscription, only one resource should be "
+                        "added or removed at a time but got {}",
+                        added_resources.size() + removed_resources.size()));
+  if (added_resources.size() == 1) {
+    ENVOY_BUG(added_resources[0].get().name() == resource_name_,
+              fmt::format("for singleton resource subscription, the added resource name "
+                          "should be the same as the subscription resource name but got {} "
+                          "!= {}",
+                          added_resources[0].get().name(), resource_name_));
+    handleAddedResource(added_resources[0]);
+  } else {
+    ENVOY_BUG(removed_resources[0] == resource_name_,
+              fmt::format("for singleton resource subscription, the removed resource name "
+                          "should be the same as the subscription resource name but got {} "
+                          "!= {}",
+                          removed_resources[0], resource_name_));
+    handleRemovedResource(removed_resources[0]);
+  }
+
+  return absl::OkStatus();
+}
+
+} // namespace LocalRateLimit
+} // namespace Common
+} // namespace Filters
+} // namespace Extensions
+} // namespace Envoy
```

Comment:
Could you move the provider singleton to the access logger limiter directory for now? Thanks.

Reply:
Hi @wbpcode , this library can be shared with be other extensions using singleton TokenBucket so this would be a right folder?

Reply:
Thoughts? @wbpcode